libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -53,603 +41,279 @@ const AP_Param::GroupInfo AP_KDECAN::var_info[] = {
-                            }
-                        } else if (res == 0) {
-                            debug_can(AP_CANManager::LOG_ERROR, "strange buffer full when stop ESC enumeration");
-                        } else {
-                            debug_can(AP_CANManager::LOG_ERROR, "error sending message to stop ESC enumeration, result %d", res);
-                        }
-                    }
-                    break;
-                }
-                case ENUMERATION_STOPPED:
-                default:
-                    debug_can(AP_CANManager::LOG_DEBUG, "something wrong happened, shouldn't be here, enumeration state: %u", enumeration_state);
-                    break;
+        const uint32_t now_ms = AP_HAL::millis();
+
+        // This should run at 400Hz
+        {
+            WITH_SEMAPHORE(_output.sem);
+            if (_output.is_new) {
+                _output.last_new_ms = now_ms;
+                _output.is_new = false;
+                memcpy(&pwm, &_output.pwm, sizeof(pwm));
+
+            } else if (_output.last_new_ms && now_ms - _output.last_new_ms > 1000) {
+                // if we haven't gotten any PWM updates for a bit, zero it
+                // out so we don't just keep sending the same values forever
+                memset(&pwm, 0, sizeof(pwm));
+                _output.last_new_ms = 0;
             }
-
-            continue;
-        }
-
-        if (!_esc_present_bitmask) {
-            hal.scheduler->delay(1000);
-            continue;
         }
 
-        // always look for received frames
-        timeout = now + LOOP_INTERVAL_US;
-
-        // check if:
-        //   - is currently sending throttle frames, OR
-        //   - there are new output values and, a throttle frame was never sent or it's no longer in CAN queue, OR
-        //   - it is time to send throttle frames again, regardless of new output values, OR
-        //   - it is time to ask for telemetry information
-        bool try_write = false;
-        if (sending_esc_num > 0 ||
-            (_new_output.load(std::memory_order_acquire) && (pwm_last_sent == 0 || now - pwm_last_sent > SET_PWM_TIMEOUT_US)) ||
-            (pwm_last_sent != 0 && (now - pwm_last_sent > SET_PWM_MIN_INTERVAL_US)) ||
-            (now - telemetry_last_request > TELEMETRY_INTERVAL_US)) {
-            // wait for write space or receive frame
-            try_write = true;
-        } else {  // don't need to send frame, choose the maximum time we'll wait for receiving a frame
-            uint64_t next_action = MIN(now + LOOP_INTERVAL_US, telemetry_last_request + TELEMETRY_INTERVAL_US);
-
-            if (pwm_last_sent != 0) {
-                next_action = MIN(next_action, pwm_last_sent + SET_PWM_MIN_INTERVAL_US);
+        uint8_t index = 0;
+        uint8_t retry = 0;
+
+        while (index < KDECAN_MAX_NUM_ESCS) {
+            if ((_init.detected_bitmask & (1 << index)) == 0) {
```

Comment:
What are the implications of the main loop updating _init.detected_bitmask while the thread is using it?

... it looks like we'll send zero to me...

Reply:
oh I see what you mean. In the old driver the common member variable _scaled_output is used to pass the data so if the mask changes then you need to set it to zero. What I'm doing is loading a zero-initialized local variable and then using a semaphore to do a memcpy to the shared memory. So yes, if it changes then it's zero.. but a better zero :)